lib/cadet/incentives/achievements.ex

```diff
@@ -21,7 +21,43 @@ defmodule Cadet.Incentives.Achievements do
     |> Repo.all()
   end
 
-  @spec upsert(map()) :: {:ok, Achievement.t()} | {:error, {:bad_request, String.t()}}
+  @doc """
+  Returns all achievements with goals and progress.
+  # To Do
+  Run elixir tests for edge cases etc.
+
+  This returns Achievement structs with prerequisites, goal association and progress maps pre-loaded.
+  """
+  def get_total_xp(course_id) when is_ecto_id(course_id) do
+    achievements =
+      Achievement
+      |> where(course_id: ^course_id)
+      |> preload([:prerequisites, goals: [goal: :progress]])
+      |> Repo.all()
+
+
+    is_goal_completed = fn (goal_item) ->
+      if (Enum.count(goal_item.goal.progress) != 0),
+        do: Enum.at(goal_item.goal.progress, 0).completed,
+        else: false
+    end
+
+    is_all_goal_completed = fn (goal_item, acc) ->
+      if is_goal_completed.(goal_item), do: {:cont, acc}, else: {:halt, 0}
+    end
+
+    acc_goals = fn (all_goals) ->
+      Enum.reduce_while(all_goals, 1, is_all_goal_completed)
+    end
+
+    acc_achievements = fn (achievement_item, acc) ->
+      acc + acc_goals.(achievement_item.goals) * achievement_item.xp
+    end
+
+    Enum.reduce(achievements, 0, acc_achievements)
+  end
+
+  @spec upsert(map()) :: {:ok, %Achievement{}} | {:error, {:bad_request, String.t()}}
```

Comment:
Did you accidentally change this line? `Achievement.t()` became `%Achievement{}`

Reply:
Unintentional, changed it back